queue.h

```diff
@@ -61,6 +63,197 @@ queue* dequeue(queue* element);
 		(element) != (listhead); \
 		(element) = (tmp))
 
+// FIXME: g++ 7.3.0, -O2, with T=route_head yields warnings.
+//        implementing QueueList as a template was meant to 
+//        i) avoid reinterpret_casts all over the code as with QUEUE_FOR_EACH
+//        ii) allow use of range based for loops.
+//        If this isn't fixed then QueueList doesn't need to be a template, T == queue.
+// In file included from defs.h:27:0,
+//                  from bend.cc:23:
+// queue.h: In instantiation of ‘const T*& QueueList<T>::ConstIterator::operator*() [with T = route_head; QueueList<T>::ConstIterator::reference = const route_head*&]’:
+// bend.cc:162:24:   required from here
+// queue.h:143:14: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
+//        return reinterpret_cast<reference>(ptr_);
+//               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+template <typename T>
+class QueueList
+{
+public:
+
+  QueueList(queue* head, int* ct) : head_{head}, ct_{ct} {}
+
+  class Iterator
+  {
+  public:
+    using iterator_category = std::bidirectional_iterator_tag;
+    using value_type = T*;
+    using difference_type = std::ptrdiff_t;
+    using pointer = T** ;
+    using reference = T* &;
+
+    Iterator() = default;
+    explicit Iterator(queue* p) : ptr_{p} {}
+
+    reference operator*()
+    {
+      return reinterpret_cast<reference>(ptr_);
+    }
+    pointer operator->()
+    {
+      return reinterpret_cast<pointer>(&ptr_);
+    }
+    Iterator& operator++()   // pre-increment
+    {
+      ptr_ = ptr_->next;
+      return *this;
+    }
+    Iterator operator++(int)   // post-increment
+    {
+      Iterator tmp = *this;
+      ++*this;
+      return tmp;
+    }
+    Iterator& operator--()   // pre-decrement
+    {
+      ptr_ = ptr_->prev;
+      return *this;
+    }
+    Iterator operator--(int)   // post-decrement
+    {
+      Iterator tmp = *this;
+      --*this;
+      return tmp;
+    }
+    bool operator==(const Iterator& other) const
+    {
+      return ptr_ == other.ptr_;
+    }
+    bool operator!=(const Iterator& other) const
+    {
+      return ptr_ != other.ptr_;
+    }
+
+  private:
+    queue* ptr_{nullptr};
+  };
+  //friend class Iterator;
+
+  class ConstIterator
```

Comment:
It's not just us, see qlist.h.

At this point class QList isn't used.  I left it in because it was a nice stepping stone in this PR, and I think it will be a nice stepping stone with a waypoint list.  Shortly after that queue.h and queue.cc go to deprecated/queue* or get removed.